tests/api/admin/controller/test_controller.py

```diff
@@ -745,6 +745,251 @@ def test_sign_out(self):
             assert None == flask.session.get("auth_type")
 
 
+class TestResetPasswordController(AdminControllerTest):
+    def test_forgot_password_get(self):
+        reset_password_ctrl = self.manager.admin_reset_password_controller
+
+        # If there is no admin with password then there is no auth providers and we should get error response
+        self.admin.password_hashed = None
+        with self.app.test_request_context("/admin/forgot_password"):
+            assert [] == reset_password_ctrl.admin_auth_providers
+
+            response = reset_password_ctrl.forgot_password()
+
+            assert response.status_code == 500
+            assert response.uri == ADMIN_AUTH_NOT_CONFIGURED.uri
+
+        # If there is only Google OAuth we should also get an error
+        create(
+            self._db,
+            ExternalIntegration,
+            protocol=ExternalIntegration.GOOGLE_OAUTH,
+            goal=ExternalIntegration.ADMIN_AUTH_GOAL,
+        )
+        with self.app.test_request_context("/admin/forgot_password"):
+            response = reset_password_ctrl.forgot_password()
+
+            assert response.status_code == 400
+            assert response.uri == ADMIN_AUTH_MECHANISM_NOT_CONFIGURED.uri
+
+        # If auth providers are set we should get forgot password page - success path
+        self.admin.password = "password"
+        with self.app.test_request_context("/admin/forgot_password"):
+            response = reset_password_ctrl.forgot_password()
+
+            assert response.status_code == 200
+            assert "Send reset password email" in response.get_data(as_text=True)
+
+        # If admin is already signed in it gets redirected since it can use regular reset password flow
+        with self.app.test_request_context("/admin/forgot_password"):
+            flask.request.form = MultiDict(
+                [
+                    ("email", self.admin.email),
+                    ("password", "password"),
+                    ("redirect", "foo"),
+                ]
+            )
+            sign_in_response = self.manager.admin_sign_in_controller.password_sign_in()
+
+            # Check that sign in is successful
+            assert sign_in_response.status_code == 302
+            assert "foo" == sign_in_response.headers["Location"]
+
+            response = reset_password_ctrl.forgot_password()
+            assert response.status_code == 302
+
+            assert "admin/web" in response.headers.get("Location")
+
+    @mock.patch("api.admin.password_admin_authentication_provider.EmailManager")
```

Comment:
Using the `patch` and a `with` context manager would make it easier to merge it into the API test fixtures changes

Reply:
Hmm, OK, I went with this approach to avoid having double `with`'s but it makes sense to change it if it will make it easier for new fixtures.

Reply:
Like [this](https://github.com/ThePalaceProject/circulation/pull/773/commits/cbf173f956c4c49f2b79ad299fb8ed3c6a7d9749)?